bftengine/src/bftengine/messages/MessageBase.cpp

```diff
@@ -46,11 +46,61 @@ void MessageBase::printLiveMessages() {
 namespace bftEngine {
 namespace impl {
 
+// static class members for diagnostics server
+std::array<std::atomic<size_t>, MsgCode::MaxMsgCodeVal> MessageBase::AliveIncomingExtrnMsgsBufs{};
+std::bitset<MsgCode::MaxMsgCodeVal> MessageBase::IncomingExtrnMsgReceivedAtLeastOnceFlags{};
+std::mutex MessageBase::messagesStatsMonitoringMutex_{};
+std::atomic<size_t> MessageBase::numIncomingExtrnMsgsBufAllocs{0};
+std::atomic<size_t> MessageBase::numIncomingExtrnMsgsBufFrees{0};
+// End of static class members for diagnostics server
+
 MessageBase::~MessageBase() {
 #ifdef DEBUG_MEMORY_MSG
   liveMessagesDebug.erase(this);
 #endif
-  if (owner_) std::free((char *)msgBody_);
+
+  // if this obj is not the owner of the buf (owner_ is false), the buf may have already be freed - hence access to it
+  // is prohibited
+  if (owner_) {
+    if (isIncomingMsg_) {
+      MsgCode::Type msg_code = static_cast<MsgCode::Type>(msgBody_->msgType);
+      updateDiagnosticsCountersOnBufRelease(msg_code);
+    }
+    std::free((char *)msgBody_);
+  }
+}
+
+void MessageBase::updateDiagnosticsCountersOnBufAlloc(MsgCode::Type msg_code) {
+  if (!IncomingExtrnMsgReceivedAtLeastOnceFlags.test(msg_code)) {
+    // here is a very rare event - first time ever receiving msg of type msg_code, hence
+    // using a lock here shouldn't affect performnce. in other places we avoid locks.
+    std::lock_guard<std::mutex> lock(messagesStatsMonitoringMutex_);
+    IncomingExtrnMsgReceivedAtLeastOnceFlags.set(msg_code);
+  }
+
+  // use "++" operator to ensure atomicity
+  numIncomingExtrnMsgsBufAllocs++;
+  AliveIncomingExtrnMsgsBufs[msg_code]++;
+}
+
+void MessageBase::updateDiagnosticsCountersOnBufRelease(MsgCode::Type msg_code) {
+  if (AliveIncomingExtrnMsgsBufs[msg_code] == 0) {
+    LOG_ERROR(GL, "Trying to dec a counter of a msg that hasn't been inserted yet, msg code: " << (msg_code));
+    return;
+  } else {
+    // use "--" operator to ensure atomicity
+    AliveIncomingExtrnMsgsBufs[msg_code]--;
+  }
+  // use "++" operator to ensure atomicity
+  numIncomingExtrnMsgsBufFrees++;
+}
+
+void MessageBase::releaseOwnership() {
```

Comment:
This is potential a new API definition.
who said its bad? its not always bad - just making it clear.

- does a call to vector::clear() x10 times produces an exception? (no, it is allowed)
- does a call to set::insert(7) x10 times produces and exception?  (no, it is allowed)
- does a call to std::unique_ptr::reset() x10 times produces an exception?  (no, it is allowed)

Its a mater of definition.
Is the api stateless or with state?
I'm not saying its bad, but it's not nesssarly the expected and eanted behavrior.
Some distructive operations are simplified by generic API which allways clearing/distroying/unregistering things without caring on the state.
Rethink if this is what you want (see tons of error in log and get lots and lots of bug fixes requsts in Jira).

Reply:
This is an existing API. it was a one line implementation in the hpp file and I moved it to the cpp file and extended it and added a safety check.

design wise I think it's a very dangerous API. it allows anyone to call it without guaranteeing another obj will take ownership of the allocated buffer, and the MessageBase obj will not release the buf's mem upon destruction.

Yulia is working on changing the msg bodies' buffer to smart ptr so I hope (we'll review the PR) it won't be a problem after she merges.